### PR TITLE
stats: add new BoolIndicator stat type

### PR DESCRIPTION
**Pull request by fredlas**
*Description*: Add the BoolIndicator stat type. A followup PR will migrate a number of gauges that only ever set 1 or 0, meaning them as true/false. Beyond adding more meaningful type distinction to stats, this will get us closer to consist logic for migrating/combining gauges during hot restart.

I chose BoolIndicator because Bool/BoolSharedPtr is not obviously something other than slightly-wrapped plain old data, and the stats classes' naming conventions would call for a function named bool(). On the other hand, I wasn't entirely sure if just Indicator would imply boolean strongly enough to make everyone happy.

This PR also converts all applicable 1-or-0 Gauges I could find to BoolIndicator.

*Risk Level*: medium
*Testing*: roughly duplicated every bit of test logic involving the word "gauge", added test verifying that gauges convert cleanly to bools across a hot restart.

#4974 #5790

### CODE REVIEW

**Comment by fredlas** [comment]
I tried migrating GrpcMux/GrpcStream's connected_state gauge over to this new type, as an example to confirm that it works in the actual code. I have left those changes out of this PR, though.

**Comment by fredlas** [comment]
I left the return type of value() as uint64_t, assuming that people could potentially have hard-to-change stuff built around that. But, taking a look at Google's use of them, I see we were actually already wanting a bool, and just doing a "> 0" to get it.

So, is it reasonable to change the return type to bool, and expect people to adapt their code?

**Comment by mattklein123** [comment]
@jmarantz are you planning on taking a pass on this one?

File: include/envoy/stats/stats.h [hunk 1]
```diff
@@ -94,5 +94,18 @@ class Gauge : public virtual Metric {
...
 
+/**
+ * A Boolean.
+ */
+class BoolIndicator : public virtual Metric {
+public:
+  virtual ~BoolIndicator() {}
+
+  virtual void set(bool value) PURE;
+  virtual uint64_t value() const PURE;
```
Comment (jmarantz) [inline include/envoy/stats/stats.h]: why do we set with a bool, but pull out the value with a uint64_t? Can we just have value() return a bool too?

Comment (fredlas) [reply include/envoy/stats/stats.h]: Sounds good, will do.

**Comment by fredlas** [comment]
No problem, there's nothing waiting on this!

Yes, there are several existing stats that will become this type. They are all among https://github.com/envoyproxy/envoy/issues/5790#issuecomment-459755480

**Comment by jmarantz** [comment]
@fredlas I think it's better to change the type now. The call-sites will need to be explicitly changed anyway from Gauge& to BoolIndicator& so that's a good time also change the value() references to expect a bool.

**Comment by fredlas** [comment]
Thanks, that makes sense.

Actually, now having gone further with the main hot restart work, I'm starting to think this change might also need to be part of the big hot-restart-breaking PR. Not sure yet, though.

**Comment by jmarantz** [comment]
I think this doesn't have to be hot-restart breaking. Just lay out the BoolIndicator impl struct exactly the same as Gauge's impl, and ensure in the impl that the bit-patterns are the same.

Leave a TODO to simplify the BoolIndicator impl once hot-restart does not use shared mem.

**Comment by fredlas** [comment]
Oh, yeah, you're totally right; as long as this change goes in before #5910 it's very straightforward. I was too far into the #5910 mindset when I wrote that!

**Comment by jmarantz** [comment]
And in fact you could write an evil unit-test to prove it, by reinterpret_cast-ing a BoolIndicator& to a Gauge& and making sure the data written to one is correctly read by the other :)

**Comment by fredlas** [comment]
I gave that test a try, and it didn't work, I'm guessing because the Gauge and BoolIndicator classes are laid out a bit differently (like BoolIndicator missing add/sub/inc/dec). Anyways, PTAL.

**Comment by jmarantz** [comment]
@fredlas I think my reinterpret_cast suggestion was a little misleading (ie wrong). Casting between the `Gauge*` and the `BoolIndicator*` doesn't make sense. It makes more sense to construct a `GaugeImpl<RawStatData>` and `BoolIndicatorImpl<RawStatData>` using the same RawStatData object declared in a test method. You should be able to then use the Gauge's write method and read from the BoolIndicator and vice versa. Does that make more sense?

**Comment by fredlas** [comment]
Yup, test added. Seeing the test pass, I think I get it, and am now largely confident that the gauge->indicator auto-conversion will work as desired. This whole stats store setup is very interesting/spooky. I didn't fully understand until now the extent to which stat name strings are important. It's like it's just a string->[raw chunk of memory] map, and the counter vs gauge classes are just syntactic sugar hiding conceptual casts of that raw memory.

In this most recent commit, I also converted every actually-a-bool gauge I could find.

PTAL!

File: docs/root/configuration/overview/v2_overview.rst [hunk 2]
```diff
@@ -590,7 +590,7 @@ Management Server has a statistics tree rooted at *control_plane.* with the foll
    :header: Name, Type, Description
    :widths: 1, 1, 2
 
-   connected_state, Gauge, A boolean (1 for connected and 0 for disconnected) that indicates the current connection state with management server
+   connected_state, BoolIndicator, Current connection state with management server
```
Comment (jmarantz) [inline docs/root/configuration/overview/v2_overview.rst]: looks like you might have accidentally deleted part of the sentence?

File: source/common/runtime/runtime_impl.h [hunk 3]
```diff
@@ -43,21 +43,21 @@ class RandomGeneratorImpl : public RandomGenerator {
...
+#define ALL_RUNTIME_STATS(BOOL_INDICATOR, COUNTER, GAUGE)                                          \
   COUNTER(load_error)                                                                              \
   COUNTER(override_dir_not_exists)                                                                 \
   COUNTER(override_dir_exists)                                                                     \
   COUNTER(load_success)                                                                            \
-  COUNTER(deprecated_feature_use)                                                                \
+  COUNTER(deprecated_feature_use)                                                                  \
   GAUGE  (num_keys)                                                                                \
-  GAUGE  (admin_overrides_active)
+  BOOL_INDICATOR  (admin_overrides_active)
```
Comment (jmarantz) [inline source/common/runtime/runtime_impl.h]: Previously the names of all the stats were aligned. That requires that you re-indent the above 6 lines. Note also that clang-format is disabled for this region.

File: source/server/server.h [hunk 4]
```diff
@@ -46,12 +46,12 @@ namespace Server {
...
  */
 // clang-format off
-#define ALL_SERVER_STATS(COUNTER, GAUGE)                                                           \
+#define ALL_SERVER_STATS(BOOL_INDICATOR, COUNTER, GAUGE)                                           \
   GAUGE(uptime)                                                                                    \
   GAUGE(concurrency)                                                                               \
   GAUGE(memory_allocated)                                                                          \
   GAUGE(memory_heap_size)                                                                          \
-  GAUGE(live)                                                                                      \
+  BOOL_INDICATOR(live)                                                                             \
```
Comment (jmarantz) [inline source/server/server.h]: probably move this to the first line of the macro, alphabetizing first by stat-type and then by stat-name.

File: test/common/stats/thread_local_store_test.cc [hunk 5]
```diff
@@ -450,6 +450,38 @@ TEST_F(StatsThreadLocalStoreTest, OverlappingScopes) {
...
+  Counter& c1 = store_->counter("samename");
+  EXPECT_EQ(&c1, &store_->counter("samename"));
+  Gauge& g1 = store_->gauge("samename");
+  EXPECT_EQ(&g1, &store_->gauge("samename"));
+  c1.add(5);
+  EXPECT_EQ(5UL, c1.value());
+  g1.add(3);
+  EXPECT_EQ(8UL, c1.value());
+
+  Gauge& g2 = store_->gauge("samename2");
```
Comment (jmarantz) [inline test/common/stats/thread_local_store_test.cc]: Is this the test you added per my suggestion? This does cover the question I wanted to answer, so thanks. However, it also locks in behavior that I think might not really be what we want once you remove hot-restart, where the underlying storage is shared independent of the type.

I think that's OK for now as you did find a way to test the functionality, but maybe we should comment that this is really an artifact of the current implementation and we might prefer to disallow Counter, Gauge, and BoolIndicator with the same name to share backing store. I think in the future we might want to make this illegal.

Of course I'm afraid of http://www.hyrumslaw.com/ here...

@mattklein123 can you share whether having a counter and gauge named the same thing and sharing backing store is intended behavior?

If we don't want to guarantee this behavior let's leave a note here that we are taking advantage of it temporarily only until the hot-restart-via-rpc PR is in.

File: test/common/upstream/resource_manager_impl_test.cc [hunk 6]
```diff
@@ -18,14 +18,14 @@ namespace Upstream {
 
 TEST(ResourceManagerImplTest, RuntimeResourceManager) {
   NiceMock<Runtime::MockLoader> runtime;
-  NiceMock<Stats::MockGauge> gauge;
+  NiceMock<Stats::MockBoolIndicator> boolIndicator;
```
Comment (jmarantz) [inline test/common/upstream/resource_manager_impl_test.cc]: bool_indicator

File: source/common/runtime/runtime_impl.h (hunk 3 above)
Comment (mattklein123) [reply source/common/runtime/runtime_impl.h]: +1 on the nit

File: docs/root/configuration/overview/v2_overview.rst (hunk 2 above)
Comment (mattklein123) [reply docs/root/configuration/overview/v2_overview.rst]: I'm fine either way, but please be consistent in lds.rst above.

File: source/server/server.cc [hunk 7]
```diff
@@ -127,6 +127,8 @@ void InstanceImpl::drainListeners() {
 
 void InstanceImpl::failHealthcheck(bool fail) {
   // We keep liveness state in shared memory so the parent process sees the same state.
+  // TODO TODO TODO OH NO, CHILD->PARENT COMMUNICATION............ i guess this is fine, just make
```
Comment (mattklein123) [inline source/server/server.cc]: nit: Please rephrase "TODO(...): ..." with an actionable statement. On the liveness topic, my recommendation here is this is one of the stats that does not transfer from the parent. (I'm not sure what your plans are there exactly in how to indicate that.)

File: test/common/stats/thread_local_store_test.cc (hunk 5 above)
Comment (mattklein123) [reply test/common/stats/thread_local_store_test.cc]: > @mattklein123 can you share whether having a counter and gauge named the same thing and sharing backing store is intended behavior?

I think I would call this an artifact of the current implementation and it's probably fine to change. So I'm fine just putting a note in somewhere to this effect.

File: docs/root/configuration/overview/v2_overview.rst (hunk 2 above)
Comment (fredlas) [reply docs/root/configuration/overview/v2_overview.rst]: These are different columns in a table - the commas are delimiters.

Matt, by consistent do you mean control_plane.connected_state vs just connected_state? If so, that difference was already there, so I left it as is - is it better to have the "control_plane." prefix, or no prefix?

File: source/common/runtime/runtime_impl.h (hunk 3 above)
Comment (fredlas) [reply source/common/runtime/runtime_impl.h]: Done.

File: source/server/server.h (hunk 4 above)
Comment (fredlas) [reply source/server/server.h]: Done.

File: test/common/stats/thread_local_store_test.cc (hunk 5 above)
Comment (fredlas) [reply test/common/stats/thread_local_store_test.cc]: Nice, note added. I am glad of this suggestion - I was surprised to learn it worked this way!

File: test/common/upstream/resource_manager_impl_test.cc (hunk 6 above)
Comment (fredlas) [reply test/common/upstream/resource_manager_impl_test.cc]: Done.

File: source/server/server.cc (hunk 7 above)
Comment (fredlas) [reply source/server/server.cc]: Ha, whoops.

Is it really ok to just let the parent-child sharing of this state go away, though? The comment sounds like it's expected that the child is the one handling the healthcheck fail command, and the use of shared memory is the only way for the parent to get this update.

(I currently have a simple stat name->transfer logic enum map for exceptions to the "sensible default" transfer logic, with the map being hardcoded with the relevant exceptions. Good at least for getting the implementation up and running for review; maybe we'll want to switch to a more robust setup, like having the behavior specified in the stat definition macros.)

**Comment by fredlas** [comment]
Merged to get whatever PR fixed the build breakage.

File: source/server/server.cc [hunk 8]
```diff
@@ -127,8 +127,9 @@ void InstanceImpl::drainListeners() {
 
 void InstanceImpl::failHealthcheck(bool fail) {
   // We keep liveness state in shared memory so the parent process sees the same state.
-  // TODO TODO TODO OH NO, CHILD->PARENT COMMUNICATION............ i guess this is fine, just make
-  // it an rpc
+  // TODO(fredlas) looks like this currently relies on shared memory for the child to convey
```
Comment (mattklein123) [inline source/server/server.cc]: This comment is not quite right. All this gauge is doing is indicating the current status of `/healthcheck/fail` and `/healthcheck/ok` admin commands, which end up failing or succeeding health checks for new HC connections. Given that the parent is told to drain, it enters effectively a perma-fail mode, and I think this stat is only going to be relevant in the primary process and should not be transferred from parent to child. Does that make more sense? I'm not sure how/if you want to modify this comment given that.

Comment (fredlas) [reply source/server/server.cc]: Ohhh, ok, thanks for the explanation. Yes, that does make more sense; sounds like nothing to worry about.

**Pull request merged**